packages/expo-device/ios/EXDevice/EXDevice.m

```diff
@@ -0,0 +1,252 @@
+}
+
+UM_EXPORT_METHOD_AS(getUserAgentAsync,
+                    getWebViewUserAgentWithResolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject)
+{
+  __weak EXDevice *weakSelf = self;
+  dispatch_async(dispatch_get_main_queue(), ^{
+    __strong EXDevice *strongSelf = weakSelf;
+    if (strongSelf) {
+      if (!strongSelf.webViewUserAgent) {
+        // We need to retain the webview because it runs an async task.
+        strongSelf->webView = [[WKWebView alloc] init];
+        
+        [strongSelf->webView evaluateJavaScript:@"window.navigator.userAgent;" completionHandler:^(id _Nullable result, NSError * _Nullable error) {
+          if (error) {
+            reject(@"ERR_CONSTANTS", error.localizedDescription, error);
+            return;
+          }
+          
+          strongSelf.webViewUserAgent = [NSString stringWithFormat:@"%@", result];
+          resolve(UMNullIfNil(strongSelf.webViewUserAgent));
+          // Destroy the webview now that it's task is complete.
+          strongSelf->webView = nil;
+        }];
+      } else {
+        resolve(UMNullIfNil(strongSelf.webViewUserAgent));
+      }
+    }
+  });
+}
+
+UM_EXPORT_METHOD_AS(getMACAddressAsync,
+                    resolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject) {
+  //some iOS privacy issues
+  NSString *address = @"02:00:00:00:00:00";
+  resolve(address);
+}
+
+UM_EXPORT_METHOD_AS(getIPAddressAsync,
+                    getIPAddressAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+  NSString *address = @"0.0.0.0";
+  struct ifaddrs *interfaces = NULL;
+  struct ifaddrs *temp_addr = NULL;
+  int success = 0;
+  // retrieve the current interfaces - returns 0 on success
+  success = getifaddrs(&interfaces);
+  if (success == 0) {
+    // Loop through linked list of interfaces
+    temp_addr = interfaces;
+    while(temp_addr != NULL) {
+      if(temp_addr->ifa_addr->sa_family == AF_INET) {
+        // Check if interface is en0 which is the wifi connection on the iPhone
+        if([[NSString stringWithUTF8String:temp_addr->ifa_name] isEqualToString:@"en0"]) {
+          // Get NSString from C String
+          address = [NSString stringWithUTF8String:inet_ntoa(((struct sockaddr_in *)temp_addr->ifa_addr)->sin_addr)];
+        }
+      }
+      temp_addr = temp_addr->ifa_next;
+    }
+  }
+  // Free memory
+  freeifaddrs(interfaces);
+  resolve(address);
+}
+
+UM_EXPORT_METHOD_AS(isPinOrFingerprintSetAsync, isPinOrFingerprintSetAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+#if TARGET_OS_TV
+  BOOL isPinOrFingerprintSet = false;
+#else
+  LAContext *context = [[LAContext alloc] init];
+  BOOL isPinOrFingerprintSet = ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]);
+#endif
+  resolve(@[[NSNumber numberWithBool:isPinOrFingerprintSet]]);
```

Comment:
@sjchmiela can they just return `BOOL`?

Reply:
BOOL, but wrapped inside NSNumber (so they’re right about that), but there is a shorthand for that – `@(isPinOrFingerPrintSet)` should do the work. Btw I’m not sure if ot would be possible to put BOOL into an array 🤔